torchx/schedulers/kubernetes_mcad_scheduler.py

```diff
@@ -0,0 +1,1121 @@
+            )
+            for name, port in role.port_map.items()
+        ],
+        volume_mounts=volume_mounts,
+        security_context=security_context,
+    )
+
+    # Get correct formatting for image secret
+    imagesecret = V1LocalObjectReference(name=image_secret)
+    return V1Pod(
+        api_version="v1",
+        kind="Pod",
+        spec=V1PodSpec(
+            containers=[container],
+            hostname=name,
+            subdomain=unique_app_id,
+            image_pull_secrets=[imagesecret],
+            restart_policy="Never",
+            service_account_name=service_account,
+            volumes=volumes,
+            node_selector=node_selector,
+        ),
+        metadata=V1ObjectMeta(
+            name=name,
+            annotations={
+                # Disable the istio sidecar as it prevents the containers from
+                # exiting once finished.
+                ANNOTATION_ISTIO_SIDECAR: "false",
+            },
+            labels={},
+            namespace=namespace,
+        ),
+    )
+
+
+def mcad_svc(svc_name: str, namespace: str, service_port: str) -> "V1Service":
+    from kubernetes.client.models import (  # noqa: F401, F811
+        V1Container,
+        V1ContainerPort,
+        V1EmptyDirVolumeSource,
+        V1EnvVar,
+        V1HostPathVolumeSource,
+        V1ObjectMeta,
+        V1PersistentVolumeClaimVolumeSource,
+        V1Pod,
+        V1PodSpec,
+        V1ResourceRequirements,
+        V1SecurityContext,
+        V1Service,
+        V1ServicePort,
+        V1ServiceSpec,
+        V1ServiceStatus,
+        V1Volume,
+        V1VolumeMount,
+    )
+
+    return V1Service(
+        api_version="v1",
+        kind="Service",
+        metadata=V1ObjectMeta(
+            name=svc_name,
+            namespace=namespace,
+        ),
+        spec=V1ServiceSpec(
+            cluster_ip="None",
+            publish_not_ready_addresses=True,
+            ports=[
+                V1ServicePort(
+                    protocol="TCP",
+                    port=int(service_port),
+                    target_port=int(service_port),
+                )
+            ],
+            selector={"appwrapper.mcad.ibm.com": svc_name},
+            session_affinity="None",
+            type="ClusterIP",
+        ),
+        status=V1ServiceStatus(
+            load_balancer={},
+        ),
+    )
+
+
+def cleanup_str(data: str) -> str:
+    """
+    Invokes ``lower`` on thes string and removes all
+    characters that do not satisfy ``[a-z0-9]`` pattern.
+    This method is mostly used to make sure kubernetes scheduler gets
+    the job name that does not violate its validation.
+    """
+    if data.startswith("-"):
+        data = data[1:]
+    pattern = r"[a-z0-9\-]"
+    return "".join(re.findall(pattern, data.lower()))
+
+
+def get_port_for_service(app: AppDef) -> str:
+    # Initialize port to default
+    port = "29500"
+
+    # look for rdzv_endpoint
+    rdzv = "--rdzv_endpoint"
```

Comment:
this seems pretty hacky -- can we instead grab the ports from the port_map in the role?

https://github.com/pytorch/torchx/blob/main/torchx/components/dist.py#L203

Reply:
Distributed PyTorch jobs fail if the ports do not match, this is related to the conversation on the mcad_svc function below.

Reply:
interesting -- can we still pull the port from the AppDef port map instead of introspecting like this? would be nice to support jobs that aren't using torch elastic

Reply:
Reran tests and was able to update to use the role port_map. Commit: 697c5aa0b348cb38b6c79447aa70eda57f5fcbcb